### ISSUE TITLE
Interface Resolution

### ISSUE THREAD

**Issue by dgocoder** [uber-go/fx#744]
I have an issue where because I am using an interface i get a missing type.

Here is the working code without fx.
```go
        gamesRepository := repository.NewMemKVS()
	gamesGateway := gateway.NewGateway()
	gamesService := services.New(gamesRepository, gamesGateway)
	gamesHandler := handlers.NewHTTPHandler(gamesService)
	router := echo.New()
	router.GET("/games/:id", gamesHandler.Get)
	router.POST("/games", gamesHandler.Create)
	router.Start(":8080")
```

I've been able to work around it with the following but am not sure if this is the correct way of going about it.
Error message: `failed to build ports.ResourceService: could not build arguments for function "main".main.func1 (/cmd/server/main.go:49): failed to build *services.Service: missing dependencies for function "/internal/core/services".New (/internal/core/services/resource.go:15): missing type: ports.ResourceGateway (did you mean *gateway.Gateway?)`

```go
         app := fx.New(
		fx.Provide(ConfigEcho),
		fx.Provide(repository.NewMemKVS),
		fx.Provide(gateway.NewGateway),
		fx.Provide(services.New,
			func(r *services.Service) ports.ResourceService { return r },
			// func(r *gateway.Gateway) ports.ResourceGateway { return r },
			func(r *repository.Respository) ports.ResourceRepository { return r }),
		fx.Provide(handlers.NewHTTPHandler),
		fx.Invoke(RegisterRoutes),
	)
	app.Run()
```

// ports file
```go
package ports

type ResourceRepository interface {
	Get(id string) (domain.Resource, error)
	Save(domain.Resource) error
}

type ResourceService interface {
	Get(id string) (domain.Resource, error)
	Create(name string) (domain.Resource, error)
}

type ResourceGateway interface {
	Send(id string) (domain.Resource, error)
}
```

// services file
```go
package services

type Service struct {
	resourceRepository ports.ResourceRepository
	resourceGateway    ports.ResourceGateway
}

func New(resourceRepository ports.ResourceRepository, resourceGateway ports.ResourceGateway) *Service {
	return &Service{
		resourceRepository: resourceRepository,
		resourceGateway:    resourceGateway,
	}
}

func (srv *Service) Get(id string) (domain.Resource, error) {
	resource, err := srv.resourceRepository.Get(id)
	if err != nil {
		return domain.Resource{}, errors.New("get failed")
	}
	return resource, nil
}

func (srv *Service) Create(name string) (domain.Resource, error) {
	game := domain.NewResource("cisco")

	if err := srv.resourceRepository.Save(game); err != nil {
		return domain.Resource{}, errors.New("create game into repository has failed")
	}
	srv.resourceGateway.Send("asdfads")
	return game, nil
}
```
//handler
```go
package handlers

type HTTPHandler struct {
	resourceService ports.ResourceService
}

func NewHTTPHandler(resourceService ports.ResourceService) *HTTPHandler {
	return &HTTPHandler{
		resourceService: resourceService,
	}
}

func (hdl *HTTPHandler) Get(c echo.Context) (err error) {
	game, err := hdl.resourceService.Get(c.Param("id"))
	fmt.Println(err)
	if err != nil {
		c.JSON(500, "string")
		return
	}

	return c.JSON(200, game)
}

func (hdl *HTTPHandler) Create(c echo.Context) (err error) {
	body := domain.Resource{}
	c.Bind(&body)

	game, err := hdl.resourceService.Create(body.Name)
	if err != nil {
		c.JSON(500, "error")
		return
	}

	return c.JSON(200, game)
}
```

**Comment by AFMiziara** [comment]
So, is this the best solution right now or `fx` does have something built-in to deal with?

**Comment by abhinav** [comment]
Hello! Yes, currently the workaround is indeed what you have in your main.go:

```go
			func(r *services.Service) ports.ResourceService { return r },
			// func(r *gateway.Gateway) ports.ResourceGateway { return r },
			func(r *repository.Respository) ports.ResourceRepository { return r }),
```

We're planning on solving this in a future version, though.
We just landed https://github.com/uber-go/dig/pull/293 in Dig which allows providing types as interfaces.
We're working on the Fx-level variant of that feature.
You can track progress in https://github.com/uber-go/fx/issues/673.

Hope this helps!

**Comment by AFMiziara** [comment]
Thanks @abhinav, I will keep up with this progress.

**Comment by AFMiziara** [comment]
I saw that #795 was merged yesterday. Is there a plan for a release including those changes?

**Comment by abhinav** [comment]
Hey @AFMiziara, yeah, we will likely release this next week.
(We avoid new releases on Fridays.)

**Comment by abhinav** [comment]
The feature has been released in [v1.15.0](https://github.com/uber-go/fx/releases/tag/v1.15.0). Thanks!
